client/src/index.js

```diff
@@ -1,30 +1,144 @@
 import ReactDOM from 'react-dom'
+import { createStore, combineReducers } from 'redux'
+import { connect, Provider } from 'react-redux'
 
-class VersionInfo extends React.Component {
-  constructor() {
-    super()
-    this.state = { versionString: '...' }
+import { setupPersist } from './persist'
+import clientVersion from './meta/version.meta'
+
+if (DEV) {
+  (async () => {
+    const response = await fetch('/api/version')
+    const apiVersion = await response.json()
+    if (clientVersion.gitBranch !== apiVersion.gitBranch
+      || clientVersion.gitHash !== apiVersion.gitHash) {
+      console.warn('Client and API versions don\'t match', clientVersion, apiVersion)
+    }
+  })()
+}
+
+function persistReducer(state = { }, action) {
+  switch (action.type) {
+    case 'LOAD_PERSIST':
+      return action.state
+
+    case 'SET_PERSIST':
+      return { ...state, ...action.data }
+
+    default:
+      return state
+  }
+}
+
+const reducer = combineReducers({
+  persist: persistReducer,
+})
+
+const store = createStore(reducer)
+
+const savePersist = setupPersist((state) => {
+  store.dispatch({ type: 'LOAD_PERSIST', state })
+})
+store.subscribe(() => {
+  savePersist(store.getState().persist || { })
+})
+
+function versionString(version) {
+  const shortHash = version.gitHash.substring(0, 8)
+  return `${shortHash} (${version.gitBranch})`
+}
+
+class NameInput extends React.Component {
+  constructor(props) {
+    super(props)
+    this.state = { name: '', nameChanged: false }
+
+    this.save = this.save.bind(this)
+    this.change = this.change.bind(this)
+  }
+
+  getName() {
+    return this.state.nameChanged ? this.state.name : this.props.name
+  }
+
+  change(e) {
+    this.setState({ name: e.target.value, nameChanged: true })
+  }
+
+  save() {
+    this.props.onSave(this.getName())
+  }
+
+  render() {
+    const name = this.getName()
+    return <>
+      <input type="text" value={name} onChange={this.change} />
+      <button type="button" onClick={this.save}>Save</button>
+    </>
+  }
+}
+
+const PersistNameInput = connect(
+  state => ({ name: state.persist.name || '' }),
+  dispatch => ({
+    onSave: (name) => {
+      dispatch({
+        type: 'SET_PERSIST',
+        data: { name },
+      })
+    },
+  }),
+)(NameInput)
+
+function NameDisplay({ name, prefix }) {
+  return <>{prefix}{name}</>
+}
+
+const PersistNameDisplay = connect(
+  state => ({ name: state.persist.name || '' }),
+)(NameDisplay)
+
+class ApiVersionInfo extends React.Component {
+  constructor(props) {
+    super(props)
+    this.state = { }
   }
 
   async componentDidMount() {
     const response = await fetch('/api/version')
     const version = await response.json()
-    const shortHash = version.gitHash.substring(0, 8)
-    const versionString = `${shortHash} (${version.gitBranch})`
-    this.setState({ versionString })
+    this.setState(version)
   }
 
   render() {
-    const { versionString } = this.state
+    const { ...state } = this.state
```

Comment:
what.

Reply:
`rEAcT/pREfEr-deStRUctUriNg`, disabled it though so could fix this